### ISSUE TITLE
Decoding damaged QR images for which I know much about in advance (Version, EC Level, payload type/size and >50% of payload itself)

### ISSUE THREAD

**Issue by MarsRover75** [zxing-cpp/zxing-cpp#828]
Hello everyone.
Is there a way to optimize detection and decoding of QR images that I know much about in advance?
What I know for sure:
1. QR is Version 3 (29x29)
2. EC Level = Q
3. Exact Payload pattern regex `^https://abc\.xy/[0-9a-zA-z]{12}$`, where abc and xy are known, so I basically have 15 out of 27 chars (+ padding)
4. The input image I get is BGR but effectively in grayscale and this can't be changed. I have to get my QRs realtime from videosource so they are 'moving targets', but with more or less predictable trajectory under normal conditions.

Right now reader doesn't return any bytes[] in result in case of `ChecksumError`.  If it did, I could at least play around with RS coding/decoding of that data.
Another maybe more straightforward way maybe is the ability to return `BitMatrix` and Mask if they were detected. This way I would be able to forcibly encode known parts (some or all), I mean 'missing' data+EC codewords into it and pass it back to decoder.

Are there any 'nice' working approaches to do it except forking the library and implementing my own additional logic?

**Comment by axxel** [comment]
Can you post an example of such a damaged QRCode symbol? If the finder patterns are (sufficiently) damaged, then the library won't find the symbol at all and all your knowledge will not help. If the symbol is detected correctly but you get a `ChecksumError`, then there are options.

 1. returning the BitMatrix is actually partly implemented in the `ZXING_EXPERIMENTAL_API` (see [here](https://github.com/zxing-cpp/zxing-cpp/blob/432df38a9526a806b0085b4275f13c390dfef8c8/core/src/Barcode.h#L179)).
 2. asking for some `bytes()` even when the checksum was wrong, has already been brought up in #786.

[...] 15 minutes later, I think I have implemented what you were looking for regarding returning some content even if a checksum error was found. :)

**Comment by MarsRover75** [comment]
> Can you post an example of such a damaged QRCode symbol?
Unfortunately I can't post the real ones here due to NDA. Maybe il''be able to forge something but it's not that easy.

I'll look into carefully your links, thanks. I already see that I'll have to implements C API & .NET wrappers for BitMatrix and maybe some more classes.

**Comment by axxel** [comment]
Since you didn't mention it, please check if [this commit](https://github.com/zxing-cpp/zxing-cpp/commit/39ebe724389913441689dc3cc56247eeeda866cb) already does what you need.

In general, I'd argue that improving the input is better than trying to improve the output. But for that, I'd need to see a sample. In case the content of a symbol is not the issue but only the background, then simply erase everything except the QRCode from the image.

**Comment by MarsRover75** [comment]
> In case the content of a symbol is not the issue but only the background

Most common issues with my images:
1. Motion blur (it's live videostream)
2. Bends and other non-linear surface distortions
3. Blinding glares & light reflections (from gloss)
so it's the content, not background.

I'm already experimenting with my Yolov8 model for detection of QR rectangles and pretrained Gs-DeblurGANv2 model for motion blur compensation, they're good but sometimes not enough.

**Comment by MarsRover75** [comment]
> > please check if [this commit](https://github.com/zxing-cpp/zxing-cpp/commit/39ebe724389913441689dc3cc56247eeeda866cb) already does what you need
> 
> I'll check this out when I have time to recompile .NET wrapper )

Yep, just to confirm: now it returns something of **variable** length, didn't have time to really dig into content.

**Comment by axxel** [comment]
> Yep, just to confirm: now it returns something of **variable** length, didn't have time to really dig into content.

Yes, and it can't really do any better: the bits stored in the symbol at the lowest level are not 'characters' but some encoded version of them and they have to be decoded first. Now if you mess up the lowest level bits, you might end up with something that can't be decoded at all or only half way. So what you get from your specific input will 'randomly' vary if you randomly change (too many) bits in your input symbol.

**Comment by MarsRover75** [comment]
Sure, that's exactly that's why I have more hopes using this approach:
`more straightforward way maybe is the ability to return BitMatrix and Mask if they were detected. This way I would be able to forcibly encode known parts (some or all), I mean 'missing' data+EC codewords into it and pass it back to decoder.
`
But
`I already see that I'll have to implement C API & .NET wrappers for BitMatrix and maybe some more classes.
`

**Comment by axxel** [comment]
@tobegreen please open a new issue and include a sample image of one of your problematic QRCodes.

**Comment by axxel** [comment]
@MarsRover75 did you pursue your approach further? Can you close this issue if you don't have any more questions?

**Comment by MarsRover75** [comment]
> @MarsRover75 did you pursue your approach further? Can you close this issue if you don't have any more questions?

Sorry, didn't have time to fully complete and test
> implement C API & .NET wrappers for BitMatrix and maybe some more classes

So I think that if there are no plans to "natively" integrate advanced techniques like this in zxing-cpp stack, maybe it's reasonable to close the issue for now.

**Comment by axxel** [comment]
> So I think if there are no plans to "natively" integrate advanced techniques like this in zxing-cpp stack (including wrappers), maybe it's reasonable to close the issue for now.

Well, access to the `BitMatrix` data will be added (see `Barcode::symbol()`) and there is currently #862 discussing the addition of a means to get access to the Mask data. But other than that, I don't see what else would be generally helpful for the "public".
